test/tools/test_tools_memories.py

```diff
@@ -58,3 +62,412 @@ def test_sopel_identifier_memory_channel_str():
+    assert memory['froMmemorY'] is True
+    assert memory['FromDict'] is False
+
+    dictionary = {'FromMemory': False, 'FromDict': True}
+    assert len(dictionary) == 2
+    assert dictionary['FromDict'] is True
+    assert dictionary['FromMemory'] is False
+
+    # __or__
+    result = memory | dictionary
+    assert len(result) == 2
+    assert result['fROmMemorY'] is False
+    assert result['fROmDicT'] is True
+
+    # __ror__
+    result = dictionary | memory
+    assert len(result) == 2
+    assert result['fROmMemorY'] is True
+    assert result['fROmDicT'] is False
+
+
+# TODO: remove condition when dropping Python 3.8
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="'|=' operator added in py3.9")
+def test_sopel_identifier_memory_ior_op():
+    memory = memories.SopelIdentifierMemory({'FromMemory': True})
+    assert len(memory) == 1
+    assert memory['FromMemory'] is True
+    assert memory['frommemory'] is True
+    assert memory['froMmemorY'] is True
+    assert 'FromDict' not in memory
+
+    other_memory = memories.SopelIdentifierMemory({
+        'FromMemory': False,
+        'FromOther': True,
+    })
+    assert len(other_memory) == 2
+    assert 'FromOther' in other_memory
+    assert other_memory['FromOther'] is True
+    assert 'fromother' in other_memory
+
+    memory |= other_memory
+    assert len(memory) == 2
+    assert memory['fROmMemorY'] is False
+    assert memory['fROmoTHEr'] is True
+    assert 'FromTuple' not in memory
+
+
+# TODO: remove condition when dropping Python 3.8
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="'|=' operator added in py3.9")
+def test_sopel_identifier_memory_ior_op_dict():
+    memory = memories.SopelIdentifierMemory({'FromMemory': True})
+    assert len(memory) == 1
+    assert memory['FromMemory'] is True
+    assert memory['frommemory'] is True
+    assert memory['froMmemorY'] is True
+    assert 'FromDict' not in memory
+
+    dictionary = {'FromDict': True}
+    assert len(dictionary) == 1
+    assert 'FromDict' in dictionary
+    assert dictionary['FromDict'] is True
+    assert 'fromdict' not in dictionary
+
+    memory |= dictionary
+    assert len(memory) == 2
+    assert memory['fROmMemorY'] is True
+    assert memory['fROmDicT'] is True
+    assert 'FromTuple' not in memory
+
+
+def test_sopel_identifier_memory_eq():
+    """Test equality checks between two `SopelIdentifierMemory` instances."""
+    memory = memories.SopelIdentifierMemory({'Foo': 'bar', 'Baz': 'Luhrmann'})
+    other_memory = memories.SopelIdentifierMemory((('Foo', 'bar'), ('Baz', 'Luhrmann')))
+
+    assert memory == other_memory
+    assert other_memory == memory  # cover our bases vis-a-vis operand precedence
+
+
+def test_sopel_identifier_memory_eq_dict():
+    """Test equality checks between `dict` and `SopelIdentifierMemory`.
+
+    Sopel's memory types are actually NOT guaranteed to compare as identical
+    with a `dict` containing the same key-value pairs.
```

Comment:
Nit: Not crazy about the wording of this docstring, because we do not in general have _"the same key-value pairs"_ because the whole point of the type is that we make `Identifier` objects. But I understand the code that it goes with, so it's not a huge deal for me.

Reply:
Repeating what I said on IRC here now that I'm at a proper computer again:

The wording still makes me uneasy because the equivalence could be misunderstood as being about the underlying `str` behind the `Identifier`. I would probably document here that the case in question IS guaranteed to NOT be an equivalence, and that's intended behavior.